src/Spice86.Core/Emulator/InterruptHandlers/VGA/VgaBios.cs

```diff
@@ -0,0 +1,1496 @@
+        }
+        address.Offset += (ushort)(character * characterHeight);
+        return address;
+    }
+
+    private SegmentedAddress GetInterruptVectorAddress(int vector) {
+        int tableOffset = 4 * vector;
+        ushort segment = _memory.UInt16[MemoryMap.InterruptVectorSegment, (ushort)(tableOffset + 2)];
+        ushort offset = _memory.UInt16[MemoryMap.InterruptVectorSegment, (ushort)tableOffset];
+        return new SegmentedAddress(segment, offset);
+    }
+
+    private CursorPosition GetCursorPosition(byte page) {
+        if (page > 7) {
+            return new CursorPosition(0, 0, 0);
+        }
+        ushort xy = _bios.GetCursorPosition(page);
+        return new CursorPosition(xy & 0xFF, xy >> 8, page);
+    }
+
+    private CharacterPlusAttribute ReadChar(CursorPosition cp) {
+        VgaMode vgaMode = _currentVgaMode;
+
+        if (vgaMode.MemoryModel != MemoryModel.Text) {
+            return ReadGraphicsCharacter(vgaMode, cp);
+        }
+
+        ushort offset = TextAddress(cp);
+        ushort value = _memory.UInt16[vgaMode.StartSegment, offset];
+        return new CharacterPlusAttribute((char)value, (byte)(value >> 8), false);
+    }
+
+    private CharacterPlusAttribute ReadGraphicsCharacter(VgaMode vgaMode, CursorPosition cursorPosition) {
+        int characterHeight = _bios.CharacterHeight;
+        if (cursorPosition.X >= _bios.ScreenColumns || characterHeight > 16) {
+            return new CharacterPlusAttribute((char)0, 0, false);
+        }
+
+        // Read cell from screen
+        GraphicsOperation operation = CreateGraphicsOperation(vgaMode);
+        operation.MemoryAction = MemoryAction.ReadByte;
+        operation.X = (ushort)(cursorPosition.X * 8);
+        operation.Y = (ushort)(cursorPosition.Y * characterHeight);
+
+        byte foregroundAttribute = 0x00;
+        const byte backgroundAttribute = 0x00;
+        byte[] lines = new byte[characterHeight];
+
+        for (byte i = 0; i < characterHeight; i++, operation.Y++) {
+            byte line = 0;
+            HandleGraphicsOperation(operation);
+            for (byte j = 0; j < 8; j++) {
+                if (operation.Pixels[j] == backgroundAttribute) {
+                    continue;
+                }
+                line |= (byte)(0x80 >> j);
+                foregroundAttribute = operation.Pixels[j];
+            }
+            lines[i] = line;
+        }
+
+        // Determine font
+        for (char character = (char)0; character < 256; character++) {
+            SegmentedAddress font = GetFontAddress(character);
+            if (MemCmp(lines, font.Segment, font.Offset, characterHeight) == 0) {
+                return new CharacterPlusAttribute(character, foregroundAttribute, false);
+            }
+        }
+
+        return new CharacterPlusAttribute((char)0, 0, false);
+    }
+
+    private int MemCmp(IReadOnlyList<byte> bytes, ushort segment, ushort offset, int length) {
```

Comment:
Could be moved to a common place maybe, not sure if it makes sense to have this reuseable?

Reply:
Since this is a port of an existing bios I tried to make it as stand-alone and independent as possible.
Nothing should have to depend on it and it shouldn't have to depend on anything.

Which is also why it reads and writes the I/O ports rather than manipulating an instance of the VideoState directly.